staging_vespalib/src/tests/state_server/state_server_test.cpp

```diff
@@ -123,6 +123,40 @@ TEST_FF("require that host is passed correctly", EchoHost(), HttpServer(0)) {
     EXPECT_EQUAL(default_result, run_cmd(make_string("curl -s http://localhost:%d/my/path -H \"Host:\"", f2.port())));
 }
 
+struct SamplingHandler : JsonGetHandler {
+    mutable vespalib::string my_host;
+    mutable vespalib::string my_path;
+    mutable std::map<vespalib::string,vespalib::string> my_params;
+    vespalib::string get(const vespalib::string &host, const vespalib::string &path,
+                         const std::map<vespalib::string,vespalib::string> &params) const override
+    {
+        my_host = host;
+        my_path = path;
+        my_params = params;
+        return "[]";
+    }
+};
+
+TEST_FF("require that request parameters can be inspected", SamplingHandler(), HttpServer(0))
+{
+    auto token = f2.repo().bind("/foo", f1);
+    EXPECT_EQUAL("[]", getPage(f2.port(), "/foo?a=b&x=y&z"));
+    EXPECT_EQUAL(f1.my_path, "/foo");
```

Comment:
To avoid ThreadSanitizer screaming at us we might need to establish some memory barriers to ensure that the `SamplingHandler` writes in the handler thread are guaranteed visible in the main test thread.

Reply:
fixed. If we really are going down this path we should make some convenience vespalib primitives to perform global (transitive) before/after synchronization. Or maybe we just need to use the Latch class for everything...